src/main/java/com/google/sps/data/LocationData.java

```diff
@@ -0,0 +1,74 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import com.google.appengine.api.datastore.EmbeddedEntity;
+import com.google.appengine.api.datastore.GeoPt;
+
+/** A class for the location information of a business. */
+public final class LocationData {
+
+  private final long id;
+  private GeoPt salePoint;
+  private String geoHash;
+  private float radius;
+
+  public LocationData(long id, GeoPt salePoint, String geoHash, float radius) {
+    this.id = id;
+    this.salePoint = salePoint;
+    this.geoHash = geoHash;
+    this.radius = radius;
+  }
+
+  public LocationData(EmbeddedEntity embeddedLocation) {
+    if (embeddedLocation == null) {
+      this.id = -1;
+      return;
```

Comment:
I'm confused why there's no compiler error on this line, isn't this function supposed to return a LocationData object?

Reply:
Yeah, maybe I am wrong and Mark can explain it to us, but from my understanding:
If there is a return type, it is void. You don't return a LocationData object, the object is instantiated when the constructor is called and the reference to the newly created object is assigned, but that is something that happens internally I guess.
Adding to these lines of code, other member variables are not assigned because automatically they are set to null or 0, only id has to be initialized because it is a final variable.

Reply:
Okay thanks for the clarification! Okay this makes sense and I understand why you do not need to assign any of the other members now.

Reply:
Yeah constructors in Java look strange. They look like a method that doesn't return anything.

This is a weird case though because you are trying to construct an object based on dynamic parameters and you don't know if they are valid. Some possibilities include verifying that EmbeddedEntity is nonnull before calling the constructor, or possibly throwing an Exception. There's much debate about whether or not constructors should throw exceptions. If you are both producing and consuming the code, you can do whatever works best for your use-case, but often times you're creating something to be consumed by someone else, in which case it may actually be a good idea to throw an exception like: ```throw new IllegalArgumentException("LocationData cannot be initialized with null EmbeddedEntity");``` , this way you communicate back what happened and why.

More details: https://stackoverflow.com/questions/6086334/is-it-good-practice-to-make-the-constructor-throw-an-exception/6086399

Reply:
I see, I like the idea of throwing an `IllegalArgumentException`. Below is an answer I gave to Ricardo about why the null check - In which case could the embeddedSaleCard be null? 

> I added it first thinking of cases when entity.getProperty("Embedded") returns null, because the property does not exist or is set as null. I added other checks to handle those cases, so right now there is no case, unless you pass it directly in your code

So yeah, I will change this to throw an Exception, so devs will need to verify before calling the constructor.